// reo-census/src/collector.js
const os = require('os');
const path = require('path');
const fs = require('fs');

/**
 * Data collector for package tracking
 * Gathers system information, package metadata, and installation context
 */
class Collector {
  constructor(config) {
    this.config = config;
  }

  /**
   * Collect all tracking data
   */
  collect() {
    try {
      const data = {
        // Activity type
        activity_type: 'npm_install',

        // Package information
        package: {
          name: this.config.getPackageName(),
          version: this.config.getPackageVersion(),
        },

        // System information
        system: {
          platform: process.platform,
          arch: process.arch,
          nodeVersion: process.version,
          osType: os.type(),
          osRelease: os.release(),
          osPlatform: os.platform(),
          cpuCount: os.cpus().length,
        },

        // Installation context
        installation: {
          timestamp: new Date().toISOString(),
          cwd: process.cwd(),
          isTopLevel: this.config.isTopLevelInstall(),
          npmVersion: process.env.npm_version || 'unknown',
          nodeEnv: process.env.NODE_ENV || 'unknown',
        },

        // Optional: Company/user information
        metadata: this.collectMetadata(),
      };

      // Add dependencies if full data mode is enabled
      if (this.config.isFullDataMode()) {
        const dependencies = this.collectDependencies();
        if (dependencies && Object.keys(dependencies).length > 0) {
          data.package.dependencies = dependencies;
        }
      }

      return data;
    } catch (error) {
      // Return minimal data if collection fails
      return {
        package: {
          name: this.config.getPackageName(),
          version: this.config.getPackageVersion(),
        },
        error: 'Collection failed',
        timestamp: new Date().toISOString(),
      };
    }
  }

  /**
   * Collect optional metadata (company info, git info, etc.)
   */
  collectMetadata() {
    const metadata = {};

    try {
      // Try to get git user info
      const gitConfig = this.getGitConfig();
      if (gitConfig.user) {
        metadata.gitUser = gitConfig.user;
      }
      
      // Email handling based on data mode
      if (gitConfig.email) {
        if (this.config.isFullDataMode()) {
          // Full mode: send full email
          metadata.email = gitConfig.email;
        } else {
          // Partial mode: only send domain
          const emailDomain = gitConfig.email.split('@')[1];
          if (emailDomain) {
            metadata.emailDomain = emailDomain;
          }
        }
      }

      // Try to detect CI/CD environment
      const ciInfo = this.detectCI();
      if (ciInfo) {
        metadata.ci = ciInfo;
      }

      // Try to get npm user info (if available)
      const npmUser = process.env.npm_config_user;
      if (npmUser) {
        metadata.npmUser = npmUser;
      }

      // Debug logging if verbose mode
      if (this.config && this.config.isVerbose()) {
        console.log('[reo-census] Git config found:', {
          user: gitConfig.user || 'not found',
          email: gitConfig.email ? 'found (domain extracted)' : 'not found',
          emailDomain: gitConfig.email ? gitConfig.email.split('@')[1] : 'none',
          gitConfigPath: path.join(require('os').homedir(), '.gitconfig')
        });
      }
    } catch (error) {
      // Silently fail - metadata is optional
      if (this.config && this.config.isVerbose()) {
        console.log('[reo-census] Metadata collection error:', error.message);
      }
    }

    return metadata;
  }

  /**
   * Read git config to get user information
   */
  getGitConfig() {
    const gitConfig = {};
    const gitConfigPath = path.join(os.homedir(), '.gitconfig');

    try {
      if (fs.existsSync(gitConfigPath)) {
        const content = fs.readFileSync(gitConfigPath, 'utf8');
        
        // Parse git config by lines - more robust than regex
        const lines = content.split('\n');
        let inUserSection = false;
        
        for (let i = 0; i < lines.length; i++) {
          const line = lines[i];
          const trimmedLine = line.trim();
          
          // Check if we're entering [user] section
          if (trimmedLine.match(/^\[user\]$/i)) {
            inUserSection = true;
            continue;
          }
          
          // Check if we're leaving the section (new section starts)
          if (trimmedLine.match(/^\[/)) {
            inUserSection = false;
            continue;
          }
          
          // If we're in [user] section, parse name and email
          if (inUserSection) {
            // Parse name = value (handle leading whitespace, quotes, spaces, comments)
            // Matches: "  name = value", "\tname=value", "name = value # comment"
            const nameMatch = line.match(/^\s*name\s*=\s*(.+?)(?:\s*[#;]|$)/i);
            if (nameMatch && !gitConfig.user) {
              let name = nameMatch[1].trim();
              name = name.replace(/^["']|["']$/g, ''); // Remove quotes
              gitConfig.user = name;
            }
            
            // Parse email = value (handle leading whitespace, quotes, spaces, comments)
            const emailMatch = line.match(/^\s*email\s*=\s*(.+?)(?:\s*[#;]|$)/i);
            if (emailMatch && !gitConfig.email) {
              let email = emailMatch[1].trim();
              email = email.replace(/^["']|["']$/g, ''); // Remove quotes
              gitConfig.email = email;
            }
          }
        }
        
        // Fallback: try regex if line-by-line parsing didn't work
        if (!gitConfig.user || !gitConfig.email) {
          const userSectionMatch = content.match(/\[user\]([\s\S]*?)(?=\[|$)/i);
          if (userSectionMatch) {
            const userSection = userSectionMatch[1];
            
            if (!gitConfig.user) {
              const nameMatch = userSection.match(/name\s*=\s*["']?([^"'\n]+)["']?/i);
              if (nameMatch) {
                gitConfig.user = nameMatch[1].trim();
              }
            }
            
            if (!gitConfig.email) {
              const emailMatch = userSection.match(/email\s*=\s*["']?([^"'\n]+)["']?/i);
              if (emailMatch) {
                gitConfig.email = emailMatch[1].trim();
              }
            }
          }
        }
      }
    } catch (error) {
      // Silently fail - git config might not exist or be unreadable
      if (this.config && this.config.isVerbose()) {
        console.log('[reo-census] Git config read error:', error.message);
        console.log('[reo-census] Git config path:', gitConfigPath);
        console.log('[reo-census] Git config exists:', fs.existsSync(gitConfigPath));
      }
    }
    
    // Always log in verbose mode for debugging
    if (this.config && this.config.isVerbose()) {
      console.log('[reo-census] Git config result:', gitConfig);
    }

    return gitConfig;
  }

  /**
   * Detect CI/CD environment
   */
  detectCI() {
    const ciEnvs = {
      CI: process.env.CI,
      GITHUB_ACTIONS: process.env.GITHUB_ACTIONS ? 'GitHub Actions' : null,
      GITLAB_CI: process.env.GITLAB_CI ? 'GitLab CI' : null,
      JENKINS: process.env.JENKINS_URL ? 'Jenkins' : null,
      TRAVIS: process.env.TRAVIS ? 'Travis CI' : null,
      CIRCLE_CI: process.env.CIRCLE_CI ? 'CircleCI' : null,
      BITBUCKET_BUILD_NUMBER: process.env.BITBUCKET_BUILD_NUMBER ? 'Bitbucket Pipelines' : null,
    };

    const detected = Object.entries(ciEnvs)
      .filter(([_, value]) => value !== null && value !== undefined)
      .map(([key, value]) => ({ [key]: value }));

    return detected.length > 0 ? detected : null;
  }

  /**
   * Collect dependencies from package.json
   * Only called in full data mode
   */
  collectDependencies() {
    try {
      const dependencies = {};
      // Get package.json from config
      const packageJson = this.config.packageJson;

      if (!packageJson) {
        return null;
      }

      // Collect dependencies
      if (packageJson.dependencies && typeof packageJson.dependencies === 'object') {
        dependencies.dependencies = packageJson.dependencies;
      }

      // Collect devDependencies
      if (packageJson.devDependencies && typeof packageJson.devDependencies === 'object') {
        dependencies.devDependencies = packageJson.devDependencies;
      }

      // Collect peerDependencies
      if (packageJson.peerDependencies && typeof packageJson.peerDependencies === 'object') {
        dependencies.peerDependencies = packageJson.peerDependencies;
      }

      // Collect optionalDependencies
      if (packageJson.optionalDependencies && typeof packageJson.optionalDependencies === 'object') {
        dependencies.optionalDependencies = packageJson.optionalDependencies;
      }

      // Return null if no dependencies found
      if (Object.keys(dependencies).length === 0) {
        return null;
      }

      return dependencies;
    } catch (error) {
      // Silently fail - dependencies are optional
      if (this.config && this.config.isVerbose()) {
        console.log('[reo-census] Dependency collection error:', error.message);
      }
      return null;
    }
  }

  /**
   * Format data for sending (can be overridden for custom formatting)
   */
  format(data) {
    return JSON.stringify(data);
  }
}

module.exports = Collector;
