// reo-census/src/config.js
const fs = require('fs');
const path = require('path');

/**
 * Configuration handler for package tracker
 * Reads environment variables and package.json settings
 */
class Config {
  constructor() {
    this.packageJsonPath = this.findPackageJson();
    this.packageJson = this.loadPackageJson();
    this.reoSettings = this.packageJson?.reoSettings || {};
  }

  /**
   * Find package.json by walking up the directory tree
   * Skips the tracker's own package.json to find the parent package
   */
  findPackageJson() {
    let currentDir = process.cwd();
    const root = path.parse(currentDir).root;
    const trackerPackageName = 'reo-census';

    if (currentDir.includes('node_modules')) {
      // Walk up until we exit node_modules
      while (currentDir !== root && currentDir.includes('node_modules')) {
        currentDir = path.dirname(currentDir);
      }

      // Non-hoisted case: reo-census is nested inside the parent package
      // e.g. node_modules/reodotdev/node_modules/reo-census
      // After walking up we land directly inside reodotdev/
      const packagePath = path.join(currentDir, 'package.json');
      if (fs.existsSync(packagePath)) {
        try {
          const pkg = JSON.parse(fs.readFileSync(packagePath, 'utf8'));
          const allDeps = {
            ...pkg.dependencies,
            ...pkg.devDependencies,
            ...pkg.peerDependencies,
            ...pkg.optionalDependencies,
          };
          if (pkg.name !== trackerPackageName && allDeps[trackerPackageName]) {
            return packagePath;
          }
        } catch (e) {}
      }

      // Hoisted case: reo-census sits at the top of node_modules
      // Scan node_modules siblings to find which package depends on reo-census
      const nodeModulesPath = path.join(currentDir, 'node_modules');
      if (fs.existsSync(nodeModulesPath)) {
        try {
          const entries = fs.readdirSync(nodeModulesPath);
          for (const entry of entries) {
            if (entry.startsWith('.')) continue;

            const candidates = [];
            if (entry.startsWith('@')) {
              try {
                const scopedEntries = fs.readdirSync(path.join(nodeModulesPath, entry));
                scopedEntries.forEach(s =>
                  candidates.push(path.join(nodeModulesPath, entry, s, 'package.json'))
                );
              } catch (e) {}
            } else {
              candidates.push(path.join(nodeModulesPath, entry, 'package.json'));
            }

            for (const candidatePath of candidates) {
              if (fs.existsSync(candidatePath)) {
                try {
                  const pkg = JSON.parse(fs.readFileSync(candidatePath, 'utf8'));
                  if (pkg.name === trackerPackageName) continue;
                  const allDeps = {
                    ...pkg.dependencies,
                    ...pkg.devDependencies,
                  };
                  if (allDeps[trackerPackageName]) {
                    return candidatePath;
                  }
                } catch (e) {}
              }
            }
          }
        } catch (e) {}
      }
    } else {
      // Not in node_modules - direct run, walk up normally
      while (currentDir !== root) {
        const packagePath = path.join(currentDir, 'package.json');
        if (fs.existsSync(packagePath)) {
          try {
            const pkg = JSON.parse(fs.readFileSync(packagePath, 'utf8'));
            if (pkg.name !== trackerPackageName) return packagePath;
          } catch (e) {}
        }
        currentDir = path.dirname(currentDir);
      }
    }

    // Fallback
    return path.join(process.cwd(), 'package.json');
  }

  /**
   * Load package.json content
   */
  loadPackageJson() {
    try {
      if (fs.existsSync(this.packageJsonPath)) {
        const content = fs.readFileSync(this.packageJsonPath, 'utf8');
        return JSON.parse(content);
      }
    } catch (error) {
      // Silently fail - package.json might not exist or be invalid
    }
    return null;
  }

  /**
   * Check if tracking is enabled
   * Respects environment variable and package.json settings
   */
  isTrackingEnabled() {
    // Check environment variable first (highest priority)
    const envOptOut = process.env.PACKAGE_TRACKER_ANALYTICS === 'false';
    if (envOptOut) {
      return false;
    }

    // Check if defaultOptIn is set to false in package.json
    if (this.reoSettings.defaultOptIn === false) {
      // In opt-in mode, check if explicitly enabled
      return process.env.PACKAGE_TRACKER_ANALYTICS === 'true';
    }

    // Default: opt-out mode (enabled by default)
    return true;
  }

  /**
   * Check if verbose logging is enabled
   */
  isVerbose() {
    return process.env.PACKAGE_TRACKER_VERBOSE === 'true';
  }

  /**
   * Get data collection mode
   * Returns 'partial' (default) or 'full'
   * Priority: env var > reoSettings.dataMode in package.json > default ('partial')
   */
  getDataMode() {
    const mode =
      process.env.PACKAGE_TRACKER_DATA_MODE ||
      this.reoSettings.dataMode ||
      'partial';
    return mode === 'full' ? 'full' : 'partial';
  }

  /**
   * Check if full data mode is enabled
   */
  isFullDataMode() {
    return this.getDataMode() === 'full';
  }

  /**
   * Get tracking endpoint URL
   * Can be configured via environment variable or package.json
   */
  getEndpointUrl() {
    return (
      process.env.PACKAGE_TRACKER_ENDPOINT ||
      this.reoSettings.endpoint ||
      'https://telemetry.reo.dev/data' // Default endpoint
    );
  }

  /**
   * Get package name from package.json
   */
  getPackageName() {
    return this.packageJson?.name || 'unknown';
  }

  /**
   * Get package version from package.json
   */
  getPackageVersion() {
    return this.packageJson?.version || 'unknown';
  }

  /**
   * Check if this is a top-level installation
   * Reo only tracks when installed as dependency or globally
   */
  isTopLevelInstall() {
    // Check if we're in node_modules (dependency) or global install
    const cwd = process.cwd();
    const isInNodeModules = cwd.includes('node_modules');
    const isGlobal = process.env.npm_config_global === 'true';
    
    // If not in node_modules and not global, it's likely top-level
    return !isInNodeModules && !isGlobal;
  }

  /**
   * Get all configuration for logging
   */
  getConfig() {
    return {
      enabled: this.isTrackingEnabled(),
      verbose: this.isVerbose(),
      endpoint: this.getEndpointUrl(),
      packageName: this.getPackageName(),
      packageVersion: this.getPackageVersion(),
      isTopLevel: this.isTopLevelInstall(),
      dataMode: this.getDataMode(),
    };
  }
}

module.exports = Config;
