// reo-census/src/sender.js
const https = require('https');
const http = require('http');

/**
 * HTTP client for sending tracking data
 * Implements retry logic, timeouts, and error handling
 */
class Sender {
  constructor(config) {
    this.config = config;
    this.endpoint = config.getEndpointUrl();
    this.maxRetries = 3;
    this.timeout = 5000; // 5 seconds per request
    this.maxTotalTimeout = 10000; // 10 seconds total maximum (including retries)
  }

  /**
   * Send tracking data to the endpoint
   * Non-blocking and best-effort (never throws)
   */
  async send(data) {
    if (!this.config.isTrackingEnabled()) {
      if (this.config.isVerbose()) {
        console.log('[reo-census] Tracking disabled');
      }
      return { success: false, reason: 'disabled' };
    }

    // Don't track top-level installations by default (like Reo)
    if (this.config.isTopLevelInstall()) {
      if (this.config.isVerbose()) {
        console.log('[reo-census] Skipping top-level installation');
      }
      return { success: false, reason: 'top-level-install' };
    }

    const payload = typeof data === 'string' ? data : JSON.stringify(data);

    if (this.config.isVerbose()) {
      console.log('[reo-census] Sending data:', payload);
    }

    try {
      // Enforce maximum total timeout to prevent hanging
      const timeoutPromise = this.sleep(this.maxTotalTimeout).then(() => {
        throw new Error('Maximum timeout exceeded');
      });

      const result = await Promise.race([
        this.sendWithRetry(payload),
        timeoutPromise,
      ]);
      return result;
    } catch (error) {
      // Never throw - this is best-effort
      if (this.config.isVerbose()) {
        console.error('[reo-census] Failed to send:', error.message);
      }
      return { success: false, error: error.message };
    }
  }

  /**
   * Send with retry logic
   */
  async sendWithRetry(payload, retryCount = 0) {
    try {
      const result = await this.makeRequest(payload);
      return { success: true, statusCode: result.statusCode };
    } catch (error) {
      if (retryCount < this.maxRetries) {
        // Exponential backoff: 1s, 2s, 4s
        const delay = Math.pow(2, retryCount) * 1000;
        await this.sleep(delay);
        return this.sendWithRetry(payload, retryCount + 1);
      }
      throw error;
    }
  }

  /**
   * Make HTTP/HTTPS request
   */
  makeRequest(payload) {
    return new Promise((resolve, reject) => {
      const url = new URL(this.endpoint);
      const isHttps = url.protocol === 'https:';
      const client = isHttps ? https : http;

      const options = {
        hostname: url.hostname,
        port: url.port || (isHttps ? 443 : 80),
        path: url.pathname + url.search,
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Content-Length': Buffer.byteLength(payload),
          'User-Agent': `reo-census/${this.config.getPackageVersion()}`,
        },
        timeout: this.timeout,
      };

      const req = client.request(options, (res) => {
        let data = '';

        res.on('data', (chunk) => {
          data += chunk;
        });

        res.on('end', () => {
          if (res.statusCode >= 200 && res.statusCode < 300) {
            resolve({ statusCode: res.statusCode, data });
          } else {
            reject(
              new Error(`Request failed with status ${res.statusCode}`)
            );
          }
        });
      });

      req.on('error', (error) => {
        reject(error);
      });

      req.on('timeout', () => {
        req.destroy();
        reject(new Error('Request timeout'));
      });

      req.write(payload);
      req.end();
    });
  }

  /**
   * Sleep utility for retry delays
   */
  sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

module.exports = Sender;
