// reo-census/src/tracker.js
const Config = require('./config');
const Collector = require('./collector');
const Sender = require('./sender');

/**
 * Main tracker orchestrator
 * Coordinates configuration, data collection, and sending
 */
class Tracker {
  constructor() {
    this.config = new Config();
    this.collector = new Collector(this.config);
    this.sender = new Sender(this.config);
  }

  /**
   * Run the tracking process
   * This is the main entry point
   */
  async track() {
    try {
      // Log tracking status if verbose
      if (this.config.isVerbose()) {
        const config = this.config.getConfig();
        console.log('[reo-census] Configuration:', JSON.stringify(config, null, 2));
      }

      // Check if tracking is enabled
      if (!this.config.isTrackingEnabled()) {
        if (this.config.isVerbose()) {
          console.log('[reo-census] Tracking is disabled');
        }
        return { success: false, reason: 'disabled' };
      }

      // Collect data
      const data = this.collector.collect();

      if (this.config.isVerbose()) {
        console.log('[reo-census] Collected data:', JSON.stringify(data, null, 2));
      }

      // Send data (non-blocking, best-effort)
      const result = await this.sender.send(data);

      // Log result if verbose
      if (this.config.isVerbose()) {
        console.log('[reo-census] Send result:', result);
      }

      return result;
    } catch (error) {
      // Never throw - this should never break installations
      if (this.config.isVerbose()) {
        console.error('[reo-census] Error:', error.message);
      }
      return { success: false, error: error.message };
    }
  }

  /**
   * Get current configuration (for debugging)
   */
  getConfig() {
    return this.config.getConfig();
  }
}

module.exports = Tracker;
